### PR TITLE
Add sources to compile result

### PR DESCRIPTION
**Pull request by fainashalts**
This PR adds a `sources` object to the `workflow-compile` result, and makes necessary changes in the `compile-common`, `compile-solidity`, `compile-vyper` and `external-compile` packages.

Of note, the vyper compilation result does not include `ast` or `legacyAST`, and external compilations will return an empty array for sources since the user does their own source management when using an external compiler.

### CODE REVIEW

File: packages/compile-solidity/run.js [hunk 1]
```diff
@@ -313,17 +315,34 @@ function detectErrors({
     );
   }
 
-  return { warnings, errors };
+  return {warnings, errors};
+}
+
+/**
+ * aggregate source information based on compiled output;
+ * this can include sources that are not contracts
```
Comment (gnidan) [inline packages/compile-solidity/run.js:323]: minor suggestion
```suggestion
 * this can include sources don't define any contracts
```

**Review by gnidan — APPROVED** [review]
Looks good, Faina!

**Review by haltman-at — COMMENTED** [review]
Thanks so much! Only reviewed this kind of quickly but LGTM!

File: packages/compile-common/src/types.ts [hunk 2]
```diff
@@ -1,7 +1,8 @@
-import { Abi, ImmutableReferences } from "@truffle/contract-schema/spec";
+import {Abi, ImmutableReferences} from "@truffle/contract-schema/spec";
 
 export type Compilation = {
-  sourceIndexes: string[]; //note: doesnot include internal sources
+  sourceIndexes: string[]; //note: does not include internal sources
+  sources: Source[];
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/compile-common/src/types.ts:5]: (Either that or you could just delete the comment, of course!)

**Pull request merged**